### PR TITLE
Fix wrong retimed detection on image sequence clip.

### PR DESCRIPTION
**Pull request by robin-ynput**
## Changelog Description

This PR addresses the bug described on https://github.com/ynput/ayon-core/issues/895 regarding retiming OTIO clip from image sequence.
* Need to conform source_range to available_range rate
* Retiming logic was when input media does not start 0 (image sequence, metadata...)
* Ensure embedded TC are treated correctly
* Need to preserve accurate discreet frame number when dealing with img sequences

### Environment prep
1. copy paste items form both **tool.poetry.dependencies** and **ayon.runtimeDependencies** sections from `ayon-core\client\pyproject.toml` into root level pyproject.toml under **tool.poetry.dev-dependencies**
2. remove `ayon-core\poetry.lock`
3. run `.\tools\manage.ps1 create-env` (windows)
> Your .poetry and .venv should be installed in root of repository

### Testing
1. Run the unit tests
run `.\tools\manage.ps1 run pytest .\tests\client\ayon_core\pipeline\editorial\` (windows)

[AY-979](https://app.clickup.com/t/6658547/AY-979)

### CODE REVIEW

File: client/ayon_core/pipeline/editorial.py
Comment (robin-ynput) [inline client/ayon_core/pipeline/editorial.py:248]: If I understood correctly the logic here, I think adding `media_in` + `source_in` was wrong but still gave false-positive cause most media available range starts from 0.

Comment (jakubjezek001) [reply client/ayon_core/pipeline/editorial.py:248]: Media sources often start with metadata for the starting timecode, so they're not always beginning from `0`.

File: client/ayon_core/pipeline/editorial.py
Comment (jakubjezek001) [inline client/ayon_core/pipeline/editorial.py:243]: well this code is kinda fun :D

File: client/ayon_core/pipeline/editorial.py
Comment (jakubjezek001) [inline client/ayon_core/pipeline/editorial.py:199]: We need to test this with actual case data, as some image sequences may start at frame 1001, while the metadata on the frame could show `00:20:36:12`. What would be the outcome in that scenario?

File: client/ayon_core/pipeline/editorial.py
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/editorial.py:259]:
```suggestion
    media_in_trimmed = source_in + offset_in
```

File: client/ayon_core/pipeline/editorial.py
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/editorial.py:188]:
```suggestion
        hasattr(otio.schema, "ImageSequenceReference")
        and isinstance(media_ref, otio.schema.ImageSequenceReference)
        and media_in != media_ref.start_frame
```

**Comment by jakubjezek001** [comment]
Before the introduction of Resolve's native OTIO, we mainly relied on two attributes:

*   The clip's source start frame as a relative range to the available media range.
*   The clip's media range as an absolute range, including the metadata timecode value.
  
With Resolve's interpretation of OTIO now in play, we consider:

*   The clip's source start frame refers to a trimmed absolute range that is available.
*   The media range remains absolute, including metadata.

We reference the media source to an absolute value with included metadata because of FFMPEG. Since we use FFMPEG for video trimming, it's crucial to have the timecode correctly defined.

First, figure out if the clip source range is relative or absolute to the media range. Next, find the relative source frame value and continue as before, making sure to clear up any confusing code that's mistakenly referring to me (old me 😄).

File: client/ayon_core/pipeline/editorial.py
Comment (robin-ynput) [reply client/ayon_core/pipeline/editorial.py:199]: I'm gonna give it a try !

**Comment by antirotor** [comment]
> ... especially since `ayon-core` doesn't have a test folder yet.

but it can - I added one for my own tests, see [feature/OP-1188_better-representation-model](https://github.com/ynput/ayon-core/tree/feature/OP-1188_better-representation-model) branch.

File: client/ayon_core/pipeline/editorial.py
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/editorial.py:191]: If we'll keep it, should this first reduce by `media_in` (I'm fully out of context BTW, never worked with OTIO).
```suggestion
        media_out = (media_out - media_in) + media_ref.start_frame + 1
```

File: client/ayon_core/pipeline/editorial.py
Comment (robin-ynput) [reply client/ayon_core/pipeline/editorial.py]: Added a unit tests and logic with image sequence with embedded TC.

File: client/ayon_core/pipeline/editorial.py
Comment (robin-ynput) [reply client/ayon_core/pipeline/editorial.py]: I believe we kind of need to apply different logic cause the computed retimed_values might be used in 2 ways:
* trimming for media inputs through ffmpeg which from my tests ignores embedded metadata and assumes input start a 0
* file copy for image sequence inputs, and we need to preserve in-out and handles as discrete frame numbers in order to remap the range to the available file list

File: client/ayon_core/pipeline/editorial.py
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/editorial.py:266]:
```suggestion
    media_out_trimmed = media_in_trimmed + (
        (
            conformed_source_range.duration.value
            * abs(time_scalar)
            + offset_out
        ) - 1
    )
```

File: tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py
Comment (jakubjezek001) [inline tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:137]: This value hint is unclear to me, could you elaborate a bit on what does it refer to?

**Comment by jakubjezek001** [comment]
The method `get_media_range_with_retimes` is used in [collect\_timeline\_instances.py](https://github.com/ynput/ayon-flame/blob/develop/client/ayon_flame/plugins/publish/collect_timeline_instances.py) on line 300 and also in [extract\_product\_resources.py](https://github.com/ynput/ayon-flame/blob/develop/client/ayon_flame/plugins/publish/extract_product_resources.py) on line 413.

  
I originally used this method to calculate how much material needs to be prepared for on-the-fly Flame transcoding. While looking into [flame\_export.py](https://github.com/ynput/ayon-flame/blob/develop/client/ayon_flame/otio/flame_export.py), which handles ocio clip data, I noticed that from line 256, the available frame range on `ImageSequenceReference` is set by frame number rather than Timecode. This confirms your assumption was correct.

When I wrote the Flame integration two years ago, I faced the challenge of using the embedded TC on the image sequence. From the code in `flame_export.py` on line 305, I can see that I was using the embedded TC to calculate the true source in/out for the clip, and later for calculating retime speed (though this equation might be incorrect).

File: client/ayon_core/pipeline/editorial.py
Comment (robin-ynput) [inline client/ayon_core/pipeline/editorial.py:281]: typo discrete

File: tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py
Comment (robin-ynput) [reply tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:137]: Totally, so embedded timecode for the img sequence is 1h meaning 90000 frames at 25fps.
Absolute soure range being `91046.625 - 91120.625` that gives us relative source range `1046.625 - 1120.625`.

So in exprimed as seconds the img sequence clip source range is:
- start_time : `1046 frames / 25 fps` = `41,84` seconds
- duration : `1120.625 - 1046.625` = `74 frames / 25 fps` = `2.94` seconds

Now as recompute duration as frames with an available range at 24.0 fps:
- start_time = `41,84` seconds x `24.0` fps = `1004.16` frames
- duration = `2.94` seconds x `24.0fps` = `70.56` frames

Finally remapping those as discrete frame numbers to the sequence:
Untitled-2024-09-19-1305

- first frame displayed is frame at duration `1004.16` = frame number `1005`
- last frame displayed is frame at duration `1004.16 + 70.56 = 1074.72` = frame number `1075`

Hopes that clarifies it.

File: tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py
Comment (robin-ynput) [inline tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:88]: todo: remove this

File: tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py
Comment (jakubjezek001) [reply tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:137]: I see, so it seems to be just a typo: `91046.625-91,120.625 25fps` should be actually `91046.625-91120.625 25fps`. The inconsistency of coma had confused me. Thank you.

Comment (jakubjezek001) [reply tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:137]: > first frame displayed is frame at duration 1004.16 = frame number 1005

The calculation makes sense for sure the only point we need to validate is if the rounding up even below **.5** will work.

**Comment by jakubjezek001** [comment]
I was doing some tests within Hiero editorial and noticed some differences in the resulting calculations. Here is the outcome before and after the fix. Since it's an old publisher, I could only copy and paste some resulting logs.

[otio_clip_publish_data_hiero.zip](https://github.com/user-attachments/files/17071832/otio_clip_publish_data_hiero.zip)

File: tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py
Comment (robin-ynput) [reply tests/client/ayon_core/pipeline/editorial/test_media_range_with_retimes.py:136]: Oh missed that, thanks for this ! It's adjusted.
For the rounding work below .5, do you want another tests for that ? 
Confirm this is properly handled by otio:
image

**Comment by robin-ynput** [comment]
> I was doing some tests within Hiero editorial and noticed some differences in the resulting calculations. Here is the outcome before and after the fix. Since it's an old publisher, I could only copy and paste some resulting logs.
> 
> [otio_clip_publish_data_hiero.zip](https://github.com/user-attachments/files/17071832/otio_clip_publish_data_hiero.zip)

Thanks for the additional tests in Hiero, I've investigated that a bit more. 
It looks like the issues comes from AYON native OTIO exporter which introduce a mismatch between the available and source ranges.

When I use Hiero native OTIO exporter, our clips are retimed as expected.
I've made an issue to fix this in `ayon-hiero`: https://github.com/ynput/ayon-hiero/issues/13

File: client/ayon_core/pipeline/editorial.py
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/editorial.py:281]:
```suggestion
        hasattr(otio.schema, "ImageSequenceReference")
        and isinstance(media_ref, otio.schema.ImageSequenceReference)
```

**Comment by iLLiCiTiT** [comment]
I'm not sure what is state of this PR?

**Comment by robin-ynput** [comment]
> For the testing data, we also need to check the retime effects (**LinearTimeWarp**, **FreezeFrame**, **TimeEffect**). Since the method name relates to retiming, this will cover the full range of functions. Great job on practically developing the unittest we really need! I'm thinking about whether we should add the test to the appropriate folder. @kalisp and @antirotor could lend a hand here, especially since `ayon-core` doesn't have a test folder yet.



> I'm not sure what is state of this PR?

I'm waiting on approval on this one.
I have addressed the feedback and added the tests, the Hiero problem that was reported will be tackled as its own in `ayon-hiero` (https://github.com/ynput/ayon-hiero/issues/13).

File: tools/manage.sh
Comment (iLLiCiTiT) [inline tools/manage.sh:160]: ???
What "Run the repository" means?

**Review by jakubjezek001 — APPROVED** [review]
Tested via pytests and also via Resolve and works perfect!

File: tools/manage.sh
Comment (iLLiCiTiT) [inline tools/manage.sh:182]: So, we're replacing `run` with `run`?

Comment (robin-ynput) [reply tools/manage.sh:182]: Technically yes, I I thought it made more sense to keep it explicit.
* first `run` is purely arbitrary from `manage.sh` sub command choices and can be changed in the future
* second `run` here is dictated by `poetry`

File: tools/manage.sh
Comment (robin-ynput) [reply tools/manage.sh:160]: I stole that from the Windows implementation in `manage.ps1`. Basically it allow you to run any command line in the context of the repository (`peotry run` underneath).
Maybe something like `Run a command in the repository environement` would make more sense ?

Comment (iLLiCiTiT) [reply tools/manage.sh:160]: I would say it is connected to https://github.com/ynput/ayon-core/pull/897/files#r1776786031 comment. The `run` is actually `poetry run` -> `Run a poetry command in the repository environement`?

Comment (robin-ynput) [reply tools/manage.sh]: This is adjusted, thanks for that.

**Review by iLLiCiTiT — COMMENTED** [review]
Code looks ok. @jakubjezek001 merge if is ready please.

**Review by iLLiCiTiT — APPROVED** [review]

File: client/ayon_core/plugins/publish/extract_otio_review.py
Comment (jakubjezek001) [inline client/ayon_core/plugins/publish/extract_otio_review.py:164]: This seems to be breaking the backward compatibility in my Hiero testing scene:

It looks like the **handle_start** and **handle_end** (originally intager) is overriden here into `opentime.RT` type.

```
Traceback (most recent call last):
  File "C:\Users\jakub\AppData\Local\Ynput\AYON\dependency_packages\ayon_2408091537_windows.zip\dependencies\pyblish\plugin.py", line 528, in __explicit_process
    runner(*args)
  File "C:\CODE\__YNPUT\ayon-core\client\ayon_core\plugins\publish\extract_otio_review.py", line 158, in process
TypeError: float() argument must be a string or a real number, not 'opentimelineio._opentime.RationalTime'
```

**Review by jakubjezek001 — CHANGES_REQUESTED** [review]
Tests in Hiero didn't go well.

File: client/ayon_core/plugins/publish/extract_otio_review.py
Comment (robin-ynput) [reply client/ayon_core/plugins/publish/extract_otio_review.py]: Thanks a lot for the double check, I completely forgot about retesting the multi-review-clips case. Totally my bad, apologies for this.

I've fixed the problem and added 2 additional unit tests to enforce those.
Also confirmed that it worked OK on Hiero (Linux) and Resolve (Windows).

**Review by jakubjezek001 — APPROVED** [review]
Fantastic work @robin-ynput ! Thanks a lot! 💫 💥

**Pull request merged**
